File: ecom-py/settings_manager.py
"""
Settings Manager - Handle application configuration with JSON persistence
"""
import json
import os
import logging
from typing import Dict, Any

logger = logging.getLogger(__name__)


class SettingsManager:
    """Manages application settings with JSON file persistence."""

    DEFAULT_SETTINGS = {
        'video': {
            'resolution_width': 1280,
            'resolution_height': 720,
            'fps': 30,
            'codec': 'mp4v'
        },
        'camera': {
            'index': 0,
            'auto_exposure': True,
            'exposure': -4,
            'gain': 0,
            'brightness': 128
        },
        'storage': {
            'video_path': 'videos',
            'database_path': 'database.db',
            'log_path': 'logs'
        },
        'app': {
            'flask_host': '127.0.0.1',
            'flask_port': 5000,
            'debug_mode': False
        },
        'compression': {
            'enabled': True,
            'codec': 'h264',       # 'h264' or 'h265'
            'crf': 23,             # 18-35 (lower = better quality)
            'preset': 'medium',    # ultrafast/fast/medium/slow
            'delete_original': True,  # Delete original after successful compression
            'priority': 'below_normal'  # 'low', 'below_normal', or 'normal'
        }
    }

    def __init__(self, settings_file='settings.json'):
        """Initialize settings manager."""
        self.settings_file = settings_file
        self.settings = self.load_settings()

    def load_settings(self) -> Dict[str, Any]:
        """Load settings from JSON file or create default if not exists."""
        if os.path.exists(self.settings_file):
            try:
                with open(self.settings_file, 'r') as f:
                    loaded_settings = json.load(f)
                    # Merge with defaults to ensure all keys exist
                    settings = self.DEFAULT_SETTINGS.copy()
                    self._deep_merge(settings, loaded_settings)
                    logger.info(f"Settings loaded from {self.settings_file}")
                    return settings
            except Exception as e:
                logger.error(f"Error loading settings: {e}, using defaults")
                return self.DEFAULT_SETTINGS.copy()
        else:
            logger.info("Settings file not found, using defaults")
            return self.DEFAULT_SETTINGS.copy()

    def save_settings(self) -> bool:
        """Save current settings to JSON file."""
        try:
            with open(self.settings_file, 'w') as f:
                json.dump(self.settings, f, indent=4)
            logger.info(f"Settings saved to {self.settings_file}")
            return True
        except Exception as e:
            logger.error(f"Error saving settings: {e}")
            return False

    def get(self, category: str, key: str, default=None):
        """Get a specific setting value."""
        return self.settings.get(category, {}).get(key, default)

    def set(self, category: str, key: str, value: Any):
        """Set a specific setting value."""
        if category not in self.settings:
            self.settings[category] = {}
        self.settings[category][key] = value

    def get_all(self) -> Dict[str, Any]:
        """Get all settings."""
        return self.settings.copy()

    def update_category(self, category: str, data: Dict[str, Any]):
        """Update an entire category of settings."""
        if category in self.settings:
            self.settings[category].update(data)
        else:
            self.settings[category] = data

    def reset_to_defaults(self):
        """Reset all settings to defaults."""
        self.settings = self.DEFAULT_SETTINGS.copy()
        logger.info("Settings reset to defaults")

    @staticmethod
    def _deep_merge(base: Dict, updates: Dict):
        """Recursively merge updates into base dictionary."""
        for key, value in updates.items():
            if key in base and isinstance(base[key], dict) and isinstance(value, dict):
                SettingsManager._deep_merge(base[key], value)
            else:
                base[key] = value

    # Convenience methods for specific settings

    def get_video_resolution(self) -> tuple:
        """Get video resolution as (width, height)."""
        return (
            self.get('video', 'resolution_width', 1280),
            self.get('video', 'resolution_height', 720)
        )

    def set_video_resolution(self, width: int, height: int):
        """Set video resolution."""
        self.set('video', 'resolution_width', width)
        self.set('video', 'resolution_height', height)

    def get_video_fps(self) -> int:
        """Get video FPS."""
        return self.get('video', 'fps', 30)

    def get_video_codec(self) -> str:
        """Get video codec."""
        return self.get('video', 'codec', 'mp4v')

    def get_camera_index(self) -> int:
        """Get camera index."""
        return self.get('camera', 'index', 0)

    def get_camera_auto_exposure(self) -> bool:
        """Get camera auto exposure setting."""
        return self.get('camera', 'auto_exposure', True)

    def get_camera_exposure(self) -> int:
        """Get camera manual exposure value (-13 to -1)."""
        return self.get('camera', 'exposure', -4)

    def get_camera_gain(self) -> int:
        """Get camera gain value (0 to 255)."""
        return self.get('camera', 'gain', 0)

    def get_camera_brightness(self) -> int:
        """Get camera brightness value (0 to 255)."""
        return self.get('camera', 'brightness', 128)

    def get_camera_exposure_settings(self) -> dict:
        """Get all camera exposure settings as a dictionary."""
        return {
            'auto_exposure': self.get_camera_auto_exposure(),
            'exposure': self.get_camera_exposure(),
            'gain': self.get_camera_gain(),
            'brightness': self.get_camera_brightness()
        }

    def get_video_storage_path(self) -> str:
        """Get video storage path (normalized for OS)."""
        path = self.get('storage', 'video_path', 'videos')
        return os.path.normpath(path) if path else 'videos'

    def get_database_path(self) -> str:
        """Get database path (normalized for OS)."""
        path = self.get('storage', 'database_path', 'database.db')
        return os.path.normpath(path) if path else 'database.db'

    def get_log_path(self) -> str:
        """Get log path (normalized for OS)."""
        path = self.get('storage', 'log_path', 'logs')
        return os.path.normpath(path) if path else 'logs'

File: ecom-mobile/lib/core/settings_manager.dart
/// Settings Manager - application configuration with JSON persistence.
///
/// Behavioral port of ecom-py/settings_manager.py: a class-level
/// DEFAULT_SETTINGS map organized by category (video/camera/storage/app/
/// compression), recursively deep-merged with any on-disk `settings.json`
/// on load, saved back as human-editable indented JSON, with typed
/// convenience getters/setters mirroring the reference naming
/// (getVideoFps, getCameraIndex, ...). SET-07/SET-08.
library;

import 'dart:convert';
import 'dart:io';

import 'package:path/path.dart' as p;

import 'logger.dart';

/// Manages application settings with JSON file persistence.
class SettingsManager {
  /// Loads settings immediately from [settingsFilePath] (deep-merged over
  /// defaults). [logger] is optional because the settings file is read
  /// before the log path is known (matching ecom-py, where settings_manager
  /// logs via a module logger configured later).
  SettingsManager(this.settingsFilePath, {Logger? logger}) : _logger = logger {
    _settings = loadSettings();
  }

  /// Full defaults, mirroring ecom-py SettingsManager.DEFAULT_SETTINGS
  /// key-for-key so a settings.json written by either app parses in both.
  static const Map<String, Map<String, Object?>> defaultSettings = {
    'video': {
      'resolution_width': 1280,
      'resolution_height': 720,
      'fps': 30,
      'codec': 'mp4v',
      // Recording bitrate cap (kbit/s) applied to the camera's Media
      // Foundation encoder. Without it the webcam writes ~15-19 Mbit/s.
      // New key for the port (ecom-py records via OpenCV, no bitrate knob).
      'recording_bitrate_kbps': 6000,
      // Burn the timestamp/label/barcode watermark into saved videos. New
      // key for the port (ecom-py always watermarks); off skips the
      // drawtext part of the post-save pass.
      'watermark_enabled': true,
    },
    'camera': {
      'index': 0,
      'auto_exposure': true,
      'exposure': -4,
      'gain': 0,
      'brightness': 128,
    },
    'storage': {
      // Public Movies collection: gallery-visible, USB-copyable, survives
      // app uninstall. main.dart write-probes this at startup and falls
      // back to the app-private 'videos' dir when it isn't writable.
      'video_path': '/storage/emulated/0/Movies/EcomTracker',
      'database_path': 'database.db',
      'log_path': 'logs',
      // Free-space floor (GB) checked before each recording start (STO-02).
      // ecom-py has no free-space guard, so this key is new to the port.
      'min_free_space_gb': 1.0,
    },
    'app': {
      'flask_host': '127.0.0.1',
      'flask_port': 5000,
      'debug_mode': false,
    },
    'compression': {
      'enabled': true,
      'codec': 'h264', // 'h264' or 'h265'
      'crf': 23, // 18-35 (lower = better quality)
      'preset': 'medium', // ultrafast/fast/medium/slow
      'delete_original': true, // delete original after successful compression
      'priority': 'below_normal', // 'low', 'below_normal', or 'normal'
    },
  };

  /// Full path of the settings.json file this manager reads/writes.
  final String settingsFilePath;

  final Logger? _logger;

  late Map<String, dynamic> _settings;

  /// Loads settings from the JSON file, deep-merged over defaults; returns
  /// a fresh copy of the defaults when the file is missing or malformed
  /// (malformed settings.json must never crash startup).
  Map<String, dynamic> loadSettings() {
    final file = File(settingsFilePath);
    if (file.existsSync()) {
      try {
        final decoded = jsonDecode(file.readAsStringSync());
        if (decoded is Map<String, dynamic>) {
          final settings = _copyDefaults();
          deepMerge(settings, decoded);
          _logger?.info('Settings loaded from $settingsFilePath');
          return settings;
        }
        _logger?.error('Error loading settings: not a JSON object, '
            'using defaults');
        return _copyDefaults();
      } catch (e) {
        _logger?.error('Error loading settings: $e, using defaults');
        return _copyDefaults();
      }
    }
    _logger?.info('Settings file not found, using defaults');
    return _copyDefaults();
  }

  /// Saves current settings to the JSON file as human-editable indented
  /// JSON (matching ecom-py's json.dump(..., indent=4)). Returns false on
  /// failure instead of throwing.
  bool saveSettings() {
    try {
      const encoder = JsonEncoder.withIndent('    ');
      File(settingsFilePath).writeAsStringSync(encoder.convert(_settings));
      _logger?.info('Settings saved to $settingsFilePath');
      return true;
    } catch (e) {
      _logger?.error('Error saving settings: $e');
      return false;
    }
  }

  /// Gets a specific setting value.
  Object? get(String category, String key, [Object? defaultValue]) {
    final cat = _settings[category];
    if (cat is Map) {
      return cat.containsKey(key) ? cat[key] : defaultValue;
    }
    return defaultValue;
  }

  /// Sets a specific setting value.
  void set(String category, String key, Object? value) {
    final cat = _settings.putIfAbsent(category, () => <String, dynamic>{});
    (cat as Map)[key] = value;
  }

  /// Returns a copy of all settings.
  Map<String, dynamic> getAll() => _deepCopy(_settings);

  /// Updates an entire category of settings (merging over existing keys).
  void updateCategory(String category, Map<String, Object?> data) {
    final existing = _settings[category];
    if (existing is Map) {
      existing.addAll(data);
    } else {
      _settings[category] = Map<String, Object?>.from(data);
    }
  }

  /// Resets all settings to defaults (in-memory only, matching ecom-py's
  /// reset_to_defaults - callers decide whether to save afterwards).
  void resetToDefaults() {
    _settings = _copyDefaults();
    _logger?.info('Settings reset to defaults');
  }

  /// Recursively merges [updates] into [base]: nested maps merge key-wise,
  /// scalar values in [updates] overwrite (exact port of ecom-py's
  /// _deep_merge).
  static void deepMerge(Map<String, dynamic> base, Map<String, dynamic> updates) {
    for (final entry in updates.entries) {
      final existing = base[entry.key];
      final value = entry.value;
      if (existing is Map<String, dynamic> && value is Map<String, dynamic>) {
        deepMerge(existing, value);
      } else if (existing is Map && value is Map) {
        final target = Map<String, dynamic>.from(existing);
        deepMerge(target, Map<String, dynamic>.from(value));
        base[entry.key] = target;
      } else {
        base[entry.key] = value;
      }
    }
  }

  static Map<String, dynamic> _copyDefaults() => _deepCopy(defaultSettings);

  static Map<String, dynamic> _deepCopy(Map source) {
    final copy = <String, dynamic>{};
    for (final entry in source.entries) {
      final value = entry.value;
      copy[entry.key as String] = value is Map ? _deepCopy(value) : value;
    }
    return copy;
  }

  // Convenience methods for specific settings (naming mirrors ecom-py's
  // get_video_fps/get_camera_index/... pattern).

  /// Gets video resolution as (width, height).
  (int, int) getVideoResolution() {
    return (
      (get('video', 'resolution_width', 1280) as num).toInt(),
      (get('video', 'resolution_height', 720) as num).toInt(),
    );
  }

  /// Sets video resolution.
  void setVideoResolution(int width, int height) {
    set('video', 'resolution_width', width);
    set('video', 'resolution_height', height);
  }

  /// Gets video FPS.
  int getVideoFps() => (get('video', 'fps', 30) as num).toInt();

  /// Gets video codec.
  String getVideoCodec() => get('video', 'codec', 'mp4v') as String;

  /// Gets camera index.
  int getCameraIndex() => (get('camera', 'index', 0) as num).toInt();

  /// Gets camera auto exposure setting.
  bool getCameraAutoExposure() =>
      get('camera', 'auto_exposure', true) as bool;

  /// Gets camera manual exposure value (-13 to -1).
  int getCameraExposure() => (get('camera', 'exposure', -4) as num).toInt();

  /// Gets camera gain value (0 to 255).
  int getCameraGain() => (get('camera', 'gain', 0) as num).toInt();

  /// Gets camera brightness value (0 to 255).
  int getCameraBrightness() =>
      (get('camera', 'brightness', 128) as num).toInt();

  /// Gets video storage path (normalized for OS, possibly relative).
  String getVideoStoragePath() {
    final path = get('storage', 'video_path', 'videos') as String?;
    return (path == null || path.isEmpty) ? 'videos' : p.normalize(path);
  }

  /// Gets database path (normalized for OS, possibly relative).
  String getDatabasePath() {
    final path = get('storage', 'database_path', 'database.db') as String?;
    return (path == null || path.isEmpty) ? 'database.db' : p.normalize(path);
  }

  /// Gets log path (normalized for OS, possibly relative).
  String getLogPath() {
    final path = get('storage', 'log_path', 'logs') as String?;
    return (path == null || path.isEmpty) ? 'logs' : p.normalize(path);
  }
}
